src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,90 @@
+        }
+
+        internal static BoundStatement Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            if (!method.Parameters.Any(x => x is SourceParameterSymbolBase param
+                                            && param.IsNullChecked))
+            {
+                return body;
+            }
+
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return (BoundStatement)rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            return AddNullChecksToBody(node);
+        }
+
+        private BoundNode AddNullChecksToBody(BoundBlock body)
+        {
+            var statementList = ArrayBuilder<BoundStatement>.GetInstance();
+            foreach (ParameterSymbol x in _method.Parameters)
+            {
+                if (x is SourceParameterSymbolBase param
+                    && param.IsNullChecked)
+                {
+                    if (!param.Type.IsValueType || param.Type.IsNullableTypeOrTypeParameter())
+                    {
+                        // PROTOTYPE : Warning or Error, see CodeGenNullCheckedParameterTests.TestNullCheckedSubstitution2
+                        continue;
+                    }
+                    var constructedIf = ConstructIfStatementForParameter(body, param);
+                    statementList.Add(constructedIf);
+                }
+            }
+            statementList.AddRange(body.Statements);
+
+            return _fact.Block(body.Locals, statementList.ToImmutableAndFree());
+        }
+
+        private BoundStatement ConstructIfStatementForParameter(BoundBlock body, SourceParameterSymbolBase parameter)
+        {
+            var loweredLeft = _fact.Parameter(parameter);
+            var loweredRight = _fact.Literal(ConstantValue.Null, parameter.Type);
+            BoundExpression paramIsNullCondition = parameter.Type.IsNullableType()
+                                                        ? BoundCall.Synthesized(body.Syntax,
+                                                            loweredLeft,
+                                                            LocalRewriter.UnsafeGetNullableMethod(
+                                                                body.Syntax,
+                                                                loweredLeft.Type as NamedTypeSymbol,
+                                                                SpecialMember.System_Nullable_T_get_HasValue,
+                                                                _method.DeclaringCompilation,
+                                                                _fact.Diagnostics))
+                                                        : (BoundExpression)_fact.ObjectEqual(loweredLeft, loweredRight);
```

Comment:
Does the `IsNullableType()` check help? We'll still need to handle an arg `T t!` where `T` is substituted at runtime with a nullable value type such as `int?`.

Perhaps this could be simplified to the following, for all cases:
```C#
_fact.ObjectEqual(
    _fact.Convert(
        _fact.Parameter(parameter),
        _fact.SpecialType(SpecialType.System_Object)),
    _fact.Literal(ConstantValue.Null, parameter.Type))
```

Reply:
This may be taking place after the `myNullable == null` to `!myNullable.HasValue` lowering has already occurred, though?

---
In reply to: [295920885](https://github.com/dotnet/roslyn/pull/36471#discussion_r295920885) [](ancestors = 295920885)

Reply:
After discussing with @RikkiGibson, we should keep the `IsNullableType()` check (or `IsNullableTypeOrTypeParameter()`) and emit the `HasValue` rather than boxing for those cases where we know at compile-time that the parameter type is a nullable value type.